core/src/main/java/org/opensearch/sql/analysis/Analyzer.java

```diff
@@ -220,6 +221,10 @@ public LogicalPlan visitFilter(Filter node, AnalysisContext context) {
     LogicalPlan child = node.getChild().get(0).accept(this, context);
     Expression condition = expressionAnalyzer.analyze(node.getCondition(), context);
 
+    if (condition instanceof OpenSearchFunctions.OpenSearchFunction) {
+      ((OpenSearchFunctions.OpenSearchFunction)condition).validateParameters(context);
+    }
```

Comment:
Any chance we can hide this behind `RelevanceFunctionResolver`. I assume it's responsible for finding function implementation that can match the given function and arguments.

Reply:
@dai-chen 
It's proving difficult to pass the `AnalysisContext` to `RelevanceFunctionResolver`. The context will have to be passed down to [BuiltinFunctionRepository.compile](https://github.com/opensearch-project/sql/blob/2.x/core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionRepository.java#L128-L138) which would affect multiple other function definitions. Another options is we could move this function call to [ExpressionAnalyzer.visitFunction](https://github.com/opensearch-project/sql/blob/2.x/core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java#L190-L197) where we could do something like:
```
  @Override
  public Expression visitFunction(Function node, AnalysisContext context) {
    FunctionName functionName = FunctionName.of(node.getFuncName());
    List<Expression> arguments =
        node.getFuncArgs().stream()
            .map(unresolvedExpression -> analyze(unresolvedExpression, context))
            .collect(Collectors.toList());
    var ret = (Expression) repository.compile(functionName, arguments);

    if (ret instanceof FunctionExpression) {
      ((FunctionExpression)ret).validateParameters(context);
    }

    return ret;
  }
```

We would need to update the `ExpressionAnalyzerTests` to reflect this change.

Reply:
I'm thinking because we have to work on https://github.com/opensearch-project/sql/issues/811 soon, we need to figure out how to move all to `opensearch` module. Could you check how Prometheus handle this in https://github.com/opensearch-project/sql/blob/2.x/prometheus/src/main/java/org/opensearch/sql/prometheus/functions/resolver/QueryRangeTableFunctionResolver.java and https://github.com/opensearch-project/sql/blob/2.x/prometheus/src/main/java/org/opensearch/sql/prometheus/storage/querybuilder/TimeRangeParametersResolver.java.

Reply:
@dai-chen Is this the example of function parameter validation in Prometheus you were thinking of?
https://github.com/opensearch-project/sql/blob/2.x/prometheus/src/main/java/org/opensearch/sql/prometheus/functions/resolver/QueryRangeTableFunctionResolver.java

Reply:
yes. posting examples I can find above.

Reply:
I think I get your point now. If I understand right, the challenge is at compile time, any string literal is valid when looking up function signature for relevancy function call. However, at runtime, we want to verify if the first/first few strings are valid OpenSearch field(s).

A quick question is any serious problem if we don't check it ahead of OpenSearch DSL execution? Or it is just difference in exception type and error message? I recall ML commands are doing the opposite. They just pass all arguments to ML client without much analysis. In this way, they don't inject any special logical to core engine.

If we really want to do this, the current approach can work for sure. My only concern is how we can move all these to `opensearch` module. I assume that has to happen soon (before 2.5 release). Option 2 seems better but actually Highlight logic needs to be moved too. Please see the proposal in TODO item no.6 here: https://github.com/opensearch-project/sql/pull/1091. Thanks!

Reply:
@dai-chen 

Another option going forward would be to alter @MaxKsyunz changes in [PR-1047](https://github.com/opensearch-project/sql/pull/1047) to pass the `AnalysisContext` down to `RelevanceFunctionResolver` rather than `FunctionProperties`. This would allow us to encapsulate this new functionality in the `RelevanceFunctionResolver` but may be quite a few minor changes.